### ISSUE TITLE
Fog distance cuts off clouds far

### ISSUE THREAD

**Issue by cryham** [RigsOfRods/rigs-of-rods#3135]
Another bug I notice, since I added fog and clouds in SR based tracks.

I think I found it to be here:
Could you just remove [this line](https://github.com/RigsOfRods/rigs-of-rods/blob/master/source/main/gfx/SkyManager.cpp#L99) and check? Or set it to that from Gui setting `App::gfx_sight_range`.
```
            // setting farclip (hacky)
            App::GetCameraManager()->GetCamera()->setFarClipDistance(fogEnd / 0.8);
```
Also BTW, I think 5000 m isn't much for max far distance. And it's not really a key setting nowadays IMO.
`        DrawGIntSlider(App::gfx_sight_range, _LC("GameSettings", "Sight range (meters)"), 100, 5000);`

To test add in `.terrn2`:
```
CaelumFogStart = 0
CaelumFogEnd = 90.0
```
and in `.os` file at end e.g.
```

	cloud_system
	{
		cloud_layer low
		{
			height 500
			coverage 0.2
		}
		cloud_layer mid
		{
			height 700
			coverage 0.6
		}
		cloud_layer high
		{
			height 1000
			coverage 0.8
		}
	}
}
```

![1657](https://github.com/RigsOfRods/rigs-of-rods/assets/720641/951b4f93-70b9-4301-b7d6-a6f6d717b657)
![1704e`](https://github.com/RigsOfRods/rigs-of-rods/assets/720641/df0f4218-caee-4eef-be3a-3605f9721f7f)

**Comment by CuriousMike56** [comment]
It appears `App::GetCameraManager()->GetCamera()->setFarClipDistance(fogEnd / 0.8);` culls far away objects. 
This is latest master:
![screenshot_2024-03-04_14-05-06_1](https://github.com/RigsOfRods/rigs-of-rods/assets/46073351/10f4bd17-260d-4cb0-b6fe-fbec42257a52)
And here's with the above line removed:
![screenshot_2024-03-04_14-05-26_1](https://github.com/RigsOfRods/rigs-of-rods/assets/46073351/29875c7b-f857-4f65-bff2-3d73a5731692)
Clouds are fully visible now, but notice the much higher triangle/batch count and subsequently lower FPS. 
![image](https://github.com/RigsOfRods/rigs-of-rods/assets/46073351/5b33834b-28b1-4409-8bd9-6cb8ece809b0)


> Also BTW, I think 5000 m isn't much for max far distance. And it's not really a key setting nowadays IMO.

Pretty sure 5000 is actually unlimited, on extremely large maps such as [Grenoble](https://forum.rigsofrods.org/resources/grenoble.399/) the entire heightmap is visible:
![screenshot_2024-03-01_14-24-59_1](https://github.com/RigsOfRods/rigs-of-rods/assets/46073351/6f7b8fad-20dc-41f4-b5b2-a15b37d5e8cf)

**Comment by cryham** [comment]
Okay. Well right now it's either fog or clouds, can't have both. So I think this should be fixed.
As for more batches, I think it'd be up to user or graphics presets to set this if it drops Fps.
Anyway 140 Fps sounds pretty high already, and over 2000 batches like way too high regardless.

**Comment by CuriousMike56** [comment]
Yeah the FPS is only because I'm on a high end PC (13700k, 4070ti), the difference will be much more drastic on the average player's lower end CPU with integrated graphics. High batch count is another pain point in RoR, Petr is aiming to fix it with his TerrnBatcher project (#3030)

I'm sure there's a proper way to fix the clouds, maybe with a patch to Caelum sources?

**Comment by CuriousMike56** [comment]
Looking at [Caelum source](https://github.com/OGRECave/ogre-caelum/blob/4bd1b0581d55896d37922483981582661ca557de/main/src/FlatCloudLayer.cpp#L70), the cloud entity is named `Caelum/FlatCloudLayer/Entity/` + random ID. Would it be possible to apply the culling to all objects except the cloud entity?

**Comment by cryham** [comment]
I don't think it's possible. This is camera far plane, it works for all.

**Comment by ohlidalp** [comment]
@paroj Can you advise please? Intuitively I'd expect some option to override the near/far clipping per render queue or maybe even per entity.

**Comment by paroj** [comment]
> Intuitively I'd expect some option to override the near/far clipping per render queue or maybe even per entity.

yes, you can set the max rendering distance per entity:
https://ogrecave.github.io/ogre/api/latest/class_ogre_1_1_movable_object.html#afe1f2a1009e3f14f36e1bcc9b1b9557e

note that using the far plane for this was wrong all along, as not rendering the objects is merely a side effect. The main purpose of the far plane is controlling the depth precision. The visibility system in Ogre is the intended tool for this.

**Comment by cryham** [comment]
I don't think we're on the same page, or maybe I don't get it, I didn't look much inside Ogre code.
Am I right here ?:
- One cloud layer here is done with a plane, which is basically 2 big triangles.
- All triangles are culled by HW on GPU. From camera far parameters also?
- Setting camera FarClipDistance, does change its frustum culling, which works well (on CPU) for many single entities, 
meshes with small triangles.
- rendering distance per entity also does the same, disable or enable rendering basing of distance.
- even if clouds had a plane with many triangels these would also get cut off and culled in HW

**Comment by paroj** [comment]
all above statements are correct. you want different CPU culling distances for different objects (cloud/ terrain - the rest). FarClipDistance can only be set globally and influences HW and CPU culling. Therefore you need to use rendering distance.

> All triangles are culled by HW on GPU (during [rasterization](https://en.wikipedia.org/wiki/Graphics_pipeline)?). From camera far parameters also?

yes, see:
https://paroj.github.io/gltut/Positioning/Tut04%20Perspective%20Projection.html#idp2563

**Comment by cryham** [comment]
Okay.
So I think the solution is still like like wrote in 1st post. To have far visibility of clouds.
But also to have less trees, adding maybe own distance to Gui or some other way to have them visible less far.
From what I know trees etc are in paged-geom and controlled by track file in .tobj like so:
`trees 0, 360,  1.30169, 1.57545,  1.0,  300, 1000,  treeD-22desMed.mesh none Des17-NileCity-veget4.png`
here `300, 1000,` is `distance1, distance2,` so it should rather be controlled by user from Gui for all tracks rather than by each track itself. I'd also BTW add a multiplier (for that 1.0 before or a new thing) for trees multiplier to reduce trees by user, from Gui too.
